src/eventHandlers/InvocationHandler.ts

```diff
@@ -25,187 +30,100 @@ export class InvocationHandler extends EventHandler<'invocationRequest', 'invoca
-        const msgCategory = `${info.name}.Invocation`;
-        function log(level: LogLevel, logCategory: LogCategory, ...args: any[]) {
-            channel.log({
-                invocationId: invocationId,
-                category: msgCategory,
-                message: format.apply(null, <[any, any[]]>args),
-                level: level,
-                logCategory,
-            });
-        }
-        function systemLog(level: LogLevel, ...args: any[]) {
-            log(level, LogCategory.System, ...args);
-        }
-        function userLog(level: LogLevel, ...args: any[]) {
-            if (isDone && !isExecutingPostInvocationHooks) {
-                let badAsyncMsg =
-                    "Warning: Unexpected call to 'log' on the context object after function execution has completed. Please check for asynchronous calls that are not awaited or calls to 'done' made before function execution completes. ";
-                badAsyncMsg += `Function name: ${info.name}. Invocation Id: ${invocationId}. `;
-                badAsyncMsg += `Learn more: ${asyncDoneLearnMoreLink}`;
-                systemLog(LogLevel.Warning, badAsyncMsg);
-            }
-            log(level, LogCategory.User, ...args);
+        // Log invocation details to ensure the invocation received by node worker
+        coreCtx.log(RpcLogLevel.Debug, RpcLogCategory.System, 'Received FunctionInvocationRequest');
+
+        const programmingModel: ProgrammingModel<unknown> = nonNullProp(channel, 'programmingModel');
+        const invocModel = programmingModel.getInvocationModel(coreCtx);
+
+        const hookData: HookData = {};
+        let { context, inputs } = await invocModel.getArguments();
+        let callback = channel.functionLoader.getCallback(functionId);
+
+        const preInvocContext: PreInvocationContext = {
+            hookData,
+            appHookData: channel.appHookData,
+            invocationContext: context,
+            functionCallback: callback,
+            inputs,
+        };
+
+        coreCtx.state = 'preInvocationHooks';
+        try {
+            await channel.executeHooks('preInvocation', preInvocContext, msg.invocationId, msgCategory);
+        } finally {
+            coreCtx.state = undefined;
         }
 
-        // Log invocation details to ensure the invocation received by node worker
-        systemLog(LogLevel.Debug, 'Received FunctionInvocationRequest');
-
-        function onDone(): void {
-            if (isDone) {
-                const message = resultIsPromise
-                    ? `Error: Choose either to return a promise or call 'done'. Do not use both in your script. Learn more: ${asyncDoneLearnMoreLink}`
-                    : "Error: 'done' has already been called. Please check your script for extraneous calls to 'done'.";
-                systemLog(LogLevel.Error, message);
-            }
-            isDone = true;
+        inputs = preInvocContext.inputs;
+        callback = preInvocContext.functionCallback;

```

Comment:
Not also overriding the context here?

Reply:
What do you mean? I'm talking about the plain `context`, which is created by the invocation model, not the `coreCtx`. This is one is created and managed by the programming model, so it makes sense if it wants to override it? How is it used by the worker?

Reply:
Yes you are correct it is the plain `context` not `coreCtx` so the worker does not create or use it directly. However, my larger point is still accurate in the sense that it has too much important information and should not be overwritten. The hooks are _not_ a part of the programming model, so it would not be a case of the programming model overriding its own context.